src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -290,6 +293,23 @@ private Path path(final Key key) {
         return Paths.get(this.dir.toString(), key.string());
     }
 
+    /**
+     * Removes empty directories.
+     * @param key Key
+     */
+    private void deleteEmptyDirs(final Optional<Key> key) {
+        try {
+            if (key.isPresent() && !key.get().string().isEmpty()
+                && this.path(key.get()).toFile().isDirectory()
+                && !Files.list(this.path(key.get())).findFirst().isPresent()) {
```

Comment:
@olenagerasimova we also need a check that the `path` we're going to remove is a child in hierarchy for `this.dir` storage root.

Reply:
@g4s8 we are checking it with the help of `Key.perent()` method